# Table of Contents

0.1 Importing Libraries

0.2 Importing Data

0.3 Exploring Original Dfs

0.4 Merging Dfs

0.5 Confirming Merged Df

0.6 Removing the Merge Flag

0.7 Exporting Merged Df

### 0.1 Importing Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

### 0.2 Importing Data

In [2]:
# Identify the file pathway to data files
path = r'C:\Users\CJ\Documents\_CJ-Stuff\Career Foundry\Data Immersion\Ach 4 - Python\2023-03 Instacart Basket Analysis'

In [3]:
# Import data from orders_checked3.csv and ords_prior_checked3.csv

In [5]:
df_orders = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_clean.pkl'))

In [6]:
df_ords_prior = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_prior_clean.pkl'))

### 0.3 Exploring Original Dfs

In [7]:
# Comparing shapes of the two dfs
df_ords_prior.shape

(32434489, 4)

In [8]:
df_orders.shape

(3421083, 7)

Merge should have 32434489 rows and 11 columns
(Columns = 4 + 7 + 1 merge field - 1 key overlap)

In [9]:
# Check the output of the two dfs
df_ords_prior.head()

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [10]:
df_orders.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  first_order_of_new_customer  
0                     NaN                         True  
1                    15.0                        False  
2                    21.0                        False  
3                    29.0                        False  
4                    28.0                        False

In [8]:
# Compare the datatypes
df_ords_prior.dtypes

order_id             int64
product_id           int32
add_to_cart_order    int16
reordered             int8
dtype: object

In [9]:
df_orders.dtypes

order_id                         int64
user_id                          int64
order_number                     int16
order_day_of_week                 int8
order_hour_of_day                 int8
days_since_prior_order         float16
first_order_of_new_customer       bool
dtype: object

The order_id column will be the key for merging the two tables.

### 0.4 Merging Dfs

In [10]:
df_merged_large = df_orders.merge(df_ords_prior, on = 'order_id', indicator = True)

### 0.5 Confirming Merged Df

In [11]:
df_merged_large.shape

(32434489, 11)

Merge has 32434489 rows and 11 columns as expected.

In [13]:
df_merged_large['_merge'].value_counts()

both          32434489
left_only            0
right_only           0
Name: _merge, dtype: int64

All rows were able to find a match in the other table.

In [12]:
df_merged_large.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_prior_order  first_order_of_new_customer  product_id  \
0                     NaN                         True         196   
1                     NaN                         True       14084   
2                     NaN                         True       12427   
3                     NaN                         True       26088   
4                     NaN                         True       26405   

   add_to_cart_order  reordered _merge  
0                  1          0   both  
1                  2          0   both  
2                  3          0   both  
3                  4          0   both  
4                  5          0   both

Table head looks as expected

In [14]:
# Checking for nulls across df
df_merged_large.isnull().sum()

order_id                             0
user_id                              0
order_number                         0
order_day_of_week                    0
order_hour_of_day                    0
days_since_prior_order         2078068
first_order_of_new_customer          0
product_id                           0
add_to_cart_order                    0
reordered                            0
_merge                               0
dtype: int64

In [17]:
df_merged_large['first_order_of_new_customer'].value_counts()

False    30356421
True      2078068
Name: first_order_of_new_customer, dtype: int64

The number of missing values for days_since_prior_order matches the number of 'True' values for first_order_of_new_customer confirming that part of the merge happened correctly.

### 0.6 Removing the Merge Flag

In [18]:
# Removing the merge flag so it doesn't interfere with future merges
df_merged_large = df_merged_large.drop(columns = ['_merge'])

In [19]:
# Confirming the correct column was droppped
df_merged_large.shape

(32434489, 10)

In [22]:
df_merged_large.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_prior_order  first_order_of_new_customer  product_id  \
0                     NaN                         True         196   
1                     NaN                         True       14084   
2                     NaN                         True       12427   
3                     NaN                         True       26088   
4                     NaN                         True       26405   

   add_to_cart_order  reordered  
0                  1          0  
1                  2          0  
2                  3          0  
3                  4          0  
4                  5          0

### 0.6 Exporting Merged Df

In [23]:
# Exporting df_merged_large after completing merge
df_merged_large.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_combined.pkl'))